include/bamit/IntervalNode.hpp

```diff
@@ -199,80 +199,78 @@ void construct_tree(std::unique_ptr<IntervalNode> & node, std::vector<Record> co
+   \param offstream_pos The resulting offstream position.
 */
-void overlap(sam_file_input_type & input,
-             std::unique_ptr<IntervalNode> const & root,
+void overlap(std::unique_ptr<IntervalNode> const & root,
              Position const & start,
              Position const & end,
-             std::vector<Record> & results)
+             std::streamoff & offstream_pos)
 {
     if (!root)
     {
         return;
     }
 
     Position cur_median = root->get_median();
-    // If the current median is overlapping the read, add all records from this node and search the left and right.
-    input.seek(root->get_file_offset());
-    std::vector<Record> current_reads{};
-
-    for (auto const & r : input | properly_mapped)
-    {
-        if (std::make_tuple(r.reference_id(), r.reference_position()) > cur_median)
-        {
-            break;
-        }
-        current_reads.emplace_back(std::make_tuple(r.reference_id().value(), r.reference_position().value()),
-                                   std::make_tuple(r.reference_id().value(), r.reference_position().value() +
-                                                                             get_length(r.cigar_sequence())),
-                                   r.file_offset());
-    }
+    // If the current median is overlapping the read, take the current offset as result and search the left further.
     if (cur_median >= start && cur_median <= end)
     {
-        results.insert(std::end(results), std::begin(current_reads), std::end(current_reads));
-        overlap(input, root->get_left_node(), start, end, results);
-        overlap(input, root->get_right_node(), start, end, results);
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_left_node(), start, end, offstream_pos);
     }
-    // If current median is to the right of the overlap, sort reads in ascending order and add all reads which
-    // start before the overlap ends.
+    // If current median is to the right of the overlap, go to the left tree.
     else if (end < cur_median)
     {
-        std::sort(current_reads.begin(), current_reads.end(), RecordComparatorStart());
-        for (auto const & r : current_reads)
-        {
-            if (r.start <= end)
-            {
-                results.push_back(r);
-            }
-            else
-            {
-                break;
-            }
-        }
-        overlap(input, root->get_left_node(), start, end, results);
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_left_node(), start, end, offstream_pos);
     }
+    // If current median is to the left of the overlap, go to the right tree.
     else if (start > cur_median)
     {
-        std::sort(current_reads.begin(), current_reads.end(), RecordComparatorEnd());
-        for (auto const & r : current_reads)
+        offstream_pos = root->get_file_offset();
+        overlap(root->get_right_node(), start, end, offstream_pos);
+    }
```

Comment:
It looks like this can be simplified a bit?
```cpp
// Set offstream_pos to the current file offset.
offstream_pos = root->get_file_offset();
// If current median is to the left of the overlap, go to the right subtree.
if (start > cur_median)
{
    overlap(root->get_right_node(), start, end, offstream_pos);
}
// Otherwise, current median must be either to the left or inside the overlap.
else
{
    overlap(root->get_left_node(), start, end, offstream_pos);
}
```

But what happened to the special case that we talked about yesterday?

Doesn't this possibly go way further left than necessary? We should think about storing start/end for the first/last record inside the node, so we can check for the current node if it overlaps the query...

Reply:
You are right, but I think changing
```cpp
if (cur_median >= start && cur_median <= end)
```
 to

```cpp
if (cur_median > start && cur_median <= end)
```
should prevent this. But maybe, I am not thinking about every problematic case.